File: src/app/core/services/restaurant-settings.service.ts
import { Injectable, computed, inject, signal } from '@angular/core';
import { Title } from '@angular/platform-browser';
import { HttpClient } from '@angular/common/http';
import { Observable, catchError, of, tap } from 'rxjs';
import {
  MediaStatusResponse,
  MediaUploadResponse,
  RestaurantSettings,
  UpdateRestaurantSettingsPayload,
} from '../../shared/models/restaurant-settings.model';

const HEX_COLOR_REGEX = /^#(?:[0-9a-fA-F]{3}){1,2}$/;

@Injectable({
  providedIn: 'root',
})
export class RestaurantSettingsService {
  private readonly http = inject(HttpClient);
  private readonly titleService = inject(Title);

  readonly settings = signal<RestaurantSettings | null>(null);
  readonly loading = signal<boolean>(false);
  readonly error = signal<string | null>(null);

  /** Readonly reactive derived signals with safe defaults */
  readonly restaurantName = computed(
    () => this.settings()?.restaurantName || 'RestaurantHub'
  );
  readonly tagline = computed(
    () => this.settings()?.tagline || 'Fresh food, delivered with care'
  );
  readonly phone = computed(() => this.settings()?.phone || '9876543210');
  readonly email = computed(
    () => this.settings()?.email || 'contact@restauranthub.com'
  );
  readonly addressLine1 = computed(
    () => this.settings()?.addressLine1 || '123 Gourmet Boulevard'
  );
  readonly addressLine2 = computed(() => this.settings()?.addressLine2 || '');
  readonly city = computed(() => this.settings()?.city || 'Mumbai');
  readonly state = computed(() => this.settings()?.state || 'Maharashtra');
  readonly pinCode = computed(() => this.settings()?.pinCode || '400001');
  readonly currencyCode = computed(
    () => this.settings()?.currencyCode || 'INR'
  );
  readonly currencySymbol = computed(
    () => this.settings()?.currencySymbol || '₹'
  );
  readonly deliveryFee = computed(() => this.settings()?.deliveryFee ?? 40);
  readonly freeDeliveryThreshold = computed(
    () => this.settings()?.freeDeliveryThreshold ?? 500
  );
  readonly estimatedDeliveryMinutes = computed(
    () => this.settings()?.estimatedDeliveryMinutes ?? 35
  );
  readonly isAcceptingOrders = computed(
    () => this.settings()?.acceptingOrders ?? true
  );
  readonly logoUrl = computed(() => this.settings()?.logoUrl || null);
  readonly heroImageUrl = computed(() => this.settings()?.heroImageUrl || null);
  readonly primaryColor = computed(
    () => this.settings()?.primaryColor || '#FF6B00'
  );
  readonly secondaryColor = computed(
    () => this.settings()?.secondaryColor || '#1E293B'
  );
  readonly gstin = computed(() => this.settings()?.gstin || null);
  readonly fssaiNumber = computed(() => this.settings()?.fssaiNumber || null);
  readonly openingTime = computed(() => this.settings()?.openingTime || null);
  readonly closingTime = computed(() => this.settings()?.closingTime || null);

  constructor() {
    this.loadSettings();
  }

  /**
   * Updates browser document title dynamically based on restaurant branding.
   */
  private updateDocumentTitle(name?: string | null): void {
    const brandName = name?.trim() || 'RestaurantHub';
    this.titleService.setTitle(`${brandName} | Online Ordering`);
  }

  /**
   * Fetches customer-safe restaurant settings from the backend.
   */
  loadSettings(): void {
    this.loading.set(true);
    this.error.set(null);

    this.http
      .get<RestaurantSettings>('/api/v1/settings')
      .pipe(
        tap((data) => {
          this.settings.set(data);
          this.loading.set(false);
          this.applyThemeColors(data.primaryColor, data.secondaryColor);
          this.updateDocumentTitle(data.restaurantName);
        }),
        catchError((err) => {
          this.loading.set(false);
          this.error.set('Failed to load restaurant settings.');
          this.updateDocumentTitle('RestaurantHub');
          return of(null);
        })
      )
      .subscribe();
  }

  /**
   * Admin mutation to update restaurant business settings and branding.
   */
  updateSettings(
    payload: UpdateRestaurantSettingsPayload
  ): Observable<RestaurantSettings> {
    return this.http
      .put<RestaurantSettings>('/api/v1/admin/settings', payload)
      .pipe(
        tap((updated) => {
          this.settings.set(updated);
          this.applyThemeColors(updated.primaryColor, updated.secondaryColor);
          this.updateDocumentTitle(updated.restaurantName);
        })
      );
  }

  /**
   * Checks if remote media upload provider is active.
   */
  checkMediaStatus(): Observable<MediaStatusResponse> {
    return this.http.get<MediaStatusResponse>('/api/v1/admin/media/status');
  }

  /**
   * Uploads an image asset to media storage with a controlled purpose (FOOD, LOGO, HERO).
   */
  uploadMedia(
    file: File,
    purpose: string = 'FOOD'
  ): Observable<MediaUploadResponse> {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('purpose', purpose);
    formData.append('folder', purpose.toLowerCase());
    return this.http.post<MediaUploadResponse>(
      '/api/v1/admin/media/images',
      formData
    );
  }

  /**
   * Safely applies validated CSS custom properties to the document root without arbitrary injection.
   */
  private applyThemeColors(
    primary?: string | null,
    secondary?: string | null
  ): void {
    if (typeof document === 'undefined' || !document.documentElement) {
      return;
    }

    if (primary && HEX_COLOR_REGEX.test(primary)) {
      document.documentElement.style.setProperty(
        '--restaurant-primary',
        primary
      );
    }
    if (secondary && HEX_COLOR_REGEX.test(secondary)) {
      document.documentElement.style.setProperty(
        '--restaurant-secondary',
        secondary
      );
    }
  }
}

File: backend/src/main/java/com/restauranthub/settings/RestaurantSettingsService.java
package com.restauranthub.settings;

import com.restauranthub.settings.dto.RestaurantSettingsResponse;
import com.restauranthub.settings.dto.UpdateRestaurantSettingsRequest;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Service governing retrieval and administrative modification of restaurant settings.
 */
@Service
public class RestaurantSettingsService {

    private final RestaurantSettingsRepository settingsRepository;

    public RestaurantSettingsService(RestaurantSettingsRepository settingsRepository) {
        this.settingsRepository = settingsRepository;
    }

    /**
     * Retrieves the authoritative RestaurantSettings entity.
     * Guarantees a valid entity is returned (creates safe default if none exists).
     */
    @Transactional(readOnly = true)
    public RestaurantSettings getActiveSettings() {
        return settingsRepository.findActiveSettings().orElseGet(this::createDefaultSettings);
    }

    /**
     * Retrieves public customer-safe response DTO.
     */
    @Transactional(readOnly = true)
    public RestaurantSettingsResponse getSettingsResponse() {
        return RestaurantSettingsResponse.fromEntity(getActiveSettings());
    }

    /**
     * Updates restaurant settings based on validated administrator payload.
     */
    @Transactional
    public RestaurantSettingsResponse updateSettings(UpdateRestaurantSettingsRequest request) {
        RestaurantSettings settings = settingsRepository.findActiveSettings().orElseGet(RestaurantSettings::new);

        settings.setRestaurantName(request.restaurantName().trim());
        settings.setTagline(request.tagline() != null ? request.tagline().trim() : null);
        settings.setPhone(request.phone().trim());
        settings.setEmail(request.email().trim().toLowerCase());
        settings.setAddressLine1(request.addressLine1().trim());
        settings.setAddressLine2(request.addressLine2() != null ? request.addressLine2().trim() : null);
        settings.setCity(request.city().trim());
        settings.setState(request.state().trim());
        settings.setPinCode(request.pinCode().trim());
        settings.setCurrencyCode(request.currencyCode().trim().toUpperCase());
        settings.setCurrencySymbol(request.currencySymbol().trim());
        settings.setDeliveryFee(request.deliveryFee());
        settings.setFreeDeliveryThreshold(request.freeDeliveryThreshold());
        settings.setEstimatedDeliveryMinutes(request.estimatedDeliveryMinutes());
        settings.setGstin(request.gstin() != null && !request.gstin().isBlank() ? request.gstin().trim().toUpperCase() : null);
        settings.setFssaiNumber(request.fssaiNumber() != null && !request.fssaiNumber().isBlank() ? request.fssaiNumber().trim() : null);
        settings.setOpeningTime(request.openingTime());
        settings.setClosingTime(request.closingTime());
        settings.setAcceptingOrders(request.acceptingOrders());
        settings.setLogoUrl(request.logoUrl() != null && !request.logoUrl().isBlank() ? request.logoUrl().trim() : null);
        settings.setHeroImageUrl(request.heroImageUrl() != null && !request.heroImageUrl().isBlank() ? request.heroImageUrl().trim() : null);
        settings.setPrimaryColor(request.primaryColor() != null && !request.primaryColor().isBlank() ? request.primaryColor().trim() : "#FF6B00");
        settings.setSecondaryColor(request.secondaryColor() != null && !request.secondaryColor().isBlank() ? request.secondaryColor().trim() : "#1E293B");
        settings.setUpdatedAt(LocalDateTime.now());

        RestaurantSettings saved = settingsRepository.save(settings);
        return RestaurantSettingsResponse.fromEntity(saved);
    }

    private RestaurantSettings createDefaultSettings() {
        RestaurantSettings defaultSettings = new RestaurantSettings();
        defaultSettings.setRestaurantName("RestaurantHub");
        defaultSettings.setTagline("Fresh food, delivered with care");
        defaultSettings.setPhone("9876543210");
        defaultSettings.setEmail("contact@restauranthub.com");
        defaultSettings.setAddressLine1("123 Gourmet Boulevard");
        defaultSettings.setAddressLine2("Near City Center");
        defaultSettings.setCity("Mumbai");
        defaultSettings.setState("Maharashtra");
        defaultSettings.setPinCode("400001");
        defaultSettings.setCurrencyCode("INR");
        defaultSettings.setCurrencySymbol("₹");
        defaultSettings.setDeliveryFee(new BigDecimal("40.00"));
        defaultSettings.setFreeDeliveryThreshold(new BigDecimal("500.00"));
        defaultSettings.setEstimatedDeliveryMinutes(35);
        defaultSettings.setAcceptingOrders(true);
        defaultSettings.setPrimaryColor("#FF6B00");
        defaultSettings.setSecondaryColor("#1E293B");
        return settingsRepository.save(defaultSettings);
    }
}
